### DISCUSSION TITLE
Environment variables exposed on the client

### DISCUSSION THREAD

**Discussion by BenjaminWFox** [vercel/next.js#23980] [Help]
# Bug report

## Describe the bug

Environment variables which do not use the `NEXT_PUBLIC_` prefix are inlined in built static files and exposed to the client despite [the documentation seeming to suggest otherwise](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser).

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:
Example repo: https://github.com/BenjaminWFox/nextjs-environment-variable-examples
Deployed app: https://nextjs-environment-variable-examples.vercel.app/

You can see that variable is inlined into the static `.html`: https://github.com/BenjaminWFox/nextjs-environment-variable-examples/blob/main/.next-demo/server/pages/index.html#L33

It appears that the intent is then to strip this out before the page is viewed on the client, but refreshing quickly or simulating a slow connection displays the variable prior to it being stripped out.

## Expected behavior

I would expect that a static page would not be allowed to use non-public environment variables. 

## Screenshots

GIF shows the non-public variable visible on refreshing the page:
![out](https://user-images.githubusercontent.com/9467046/97218996-a5398980-1786-11eb-93bf-a6ed300c98cc.gif)

## System information

- Browser [chrome]
- Version of Next.js: [9.5.5]
- Version of Node.js: [12.13.1 locally]

## Additional context

I saw this while trying to better understand the behavior of environment variables generally, so not sure if there was something else I might be doing wrong.

**Comment by merodiro** [comment]
the reason it appears on the first render and then disappears when the page loads is that the first render is from the server/generated HTML where it has these all env variables available but it disappears when react hydrate finishes because the env isn't available in the client bundle

**Comment by BenjaminWFox** [comment]
I get that, but considering the docs it would be easy for someone to make a mistaken assumption about how safe their environment variables actually were. They are exposed to the browser, albeit briefly.

> By default all environment variables loaded through .env.local are only available in the Node.js environment, meaning they won't be exposed to the browser.

So if not a bug maybe just a 'needs clarification'?

**Reply by robertwbradford** [reply]
The wording in this example is also incorrect: https://github.com/vercel/next.js/blob/8d0561ebf866a7bf61b74574902e557a1c85c94a/examples/environment-variables/pages/index.js#L107-L108

**Comment by BenjaminWFox** [comment]
In a similar vein, SSG files that use `process.env.*` values can use any value from the `.env`/`.env.local` file without requiring the `NEXT_PUBLIC_` prefix.

I feels like I should get an error when building in this case, and be required to prefix any environment variables for SSG files with `NEXT_PUBLIC_`.

**Comment by scottk2** [comment]
I get this same issue. No env properties should be exposed to the browser unless they are being referenced by client code. So with Next, that's any component page etc. But it shouldn't include api pages. They are meant to be treated like pure backend and never seen by the client.  I have client and a server config files - the server ones are only referenced from API/backend code. Client config can be seen by both. I know if I were running my api on express, then this wouldn't happen. Its a a pretty big issue and a maaaaaajjjjjjjjor security concern.
Might have to ditch serverless.
